### ISSUE TITLE
Connection to SMTP server fails when TLS is enabled

### ISSUE THREAD

**Issue by whalehub** [Ylianst/MeshCentral#443]
I receive the following error message in MeshCentral's output when I enable TLS while using an SMTP server that's capable of handling TLS connections on port 587:

```
Aug 23 05:56:51 debian node[36246]: SMTP mail server mail.gandi.net failed: {"library":"SSL routines","function":"ssl3_get_record","reason":"wrong version number","code":"ESOCKET","command":"CONN"}
```

This is the relevant snippet from my `config.json file`:

```
   "smtp": {
   "host": "mail.gandi.net",
   "port": 587,
   "from": "<redacted>",
   "user": "<redacted>",
   "pass": "<redacted>",
   "tls": true
  },
```

The connection works fine fine when I change `tls` to `false`, but I'd prefer to use TLS. If I'm not mistaken, it looks like MeshCentral attempts to connect to the SMTP server via SSL3 instead of TLS.

**Comment by Ylianst** [comment]
Good report, I will check this.

**Comment by whalehub** [comment]
@Ylianst I just upgraded to MeshCentral v0.4.0-k and restarted MeshCentral and I'm still receiving the following error message in the output:

```
Aug 31 09:08:17 debian node[48527]: SMTP mail server mail.gandi.net failed: {"library":"SSL routines","function":"ssl3_get_record","reason":"wrong version number","code":"ESOCKET","command":"CONN"}
```

**Comment by whalehub** [comment]
@Ylianst I just upgraded to MeshCentral v0.4.0-k and restarted MeshCentral and I'm still receiving the following error message in the output:

```
Aug 31 09:08:17 debian node[48527]: SMTP mail server mail.gandi.net failed: {"library":"SSL routines","function":"ssl3_get_record","reason":"wrong version number","code":"ESOCKET","command":"CONN"}
```

I had a look at the file you mentioned and noticed that nodemailer is being told to use SSL2/3 with the piece of code `secureProtocol: 'SSLv23_method'`. If I'm not mistaken, it should say `secureProtocol: 'TLSv1_2_method'` instead (see https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options for reference).

**Comment by whalehub** [comment]
@Ylianst Any update on this?

**Comment by OutbackMatt** [comment]
My Mailserver is set to use TLSv1.2 or TLSv1.3 only, and meshcentral can't connect to my server using TLS as version is too low.

I agree that the options are set incorrectly.

**Comment by OutbackMatt** [comment]
A fix for my server

./node_modules/meshcentral/meshmail.js
Line 35 change to 
    var options = { host: parent.config.smtp.host, secure: (parent.config.smtp.tls == true) };

** This just lets the defaults work, and it negotiates whatever is available


And then the important bit for me was to realise that Secure = True means SSL connections (security first, then make contact), and that Secure = False means StartTLS connection

gMail and office365 (like most mail servers) accept port 587 StartTLS, for submission of email and NodeMailer considers this NOT secure. Once I set TLS = False in my 'config.json' then I can see mail connecting via START TLS using TLSv1.3 which is really awesome and what I wanted.

**Comment by Ylianst** [comment]
Ha! Thanks. Just published MeshCentral v0.4.0-u with your "options". You are correct about the TLS option, I have it set to "false" myself and the START TLS happens just as you describe.

**Comment by OutbackMatt** [comment]
Thanks - That seems to work
Have updated to the new version and it works fine.

**Comment by whalehub** [comment]
@Ylianst Thanks for the fix.

**Comment by whalehub** [comment]
@OutbackMatt I did a bit more testing with this setting today and discovered that setting `tls` to `true` will only work if and only if the SMTP server supports Implicit TLS (smtps) on port 465. This is the preferred option since STARTTLS on port 587 is opportunistic and allows a MITM to downgrade the connection to an unencrypted one whereas Implicit TLS on port 465 mandates the use of encryption.

```
   "smtp": {
   "host": "mail.gandi.net",
   "port": 465,
   "from": "<redacted>",
   "user": "<redacted>",
   "pass": "<redacted>",
   "tls": true
  },
```

@Ylianst I think the email setup example in MeshCentral's [User Guide](http://info.meshcentral.com/downloads/MeshCentral2/MeshCentral2UserGuide-0.2.6.pdf) should be updated with this information and port 25 should be replaced with port 587 since the former has been reserved for connections between MTAs for well over a decade by now.

**Comment by OutbackMatt** [comment]
My Server allows both port 465 SMTPS and StartTLS SMTP on port 587
Settings as I described above was required at the time I posted.

Checking today, I see that port 465 SMTPS and using TLS = true
 does in fact work as you describe.

MITM is possible with both SMTPS and StartTLS. The way to beat MITM would be to use named certs, and require that your server when connecting to others validates certs. Also using DNSSEC and DANE etc helps

Downgrade with StartTLS isn't a problem unless you allow lower levels of encryption than TLSv1.2 (ie TLSv1.1 or TLSv1,0 or SSLv3.0), OR no encryption. StartTLS can be forced on many servers (including hMailserver that I use), and by gmail, office365 etc), so it can't ever be downgraded to an un-encrypted connection.
